src/main/java/org/zalando/riptide/TypeInference.java

```diff
@@ -0,0 +1,180 @@
+                .orElseThrow(() -> new IllegalStateException("Could not extract generic types from [" + function.getClass() + "]"));
+    }
+
+    private static boolean isStatic(final Method method) {
+        return (method.getModifiers() & Modifier.STATIC) != 0;
+    }
+
+    private static <T, R> FunctionInfo<T, R> getFunctionInfoForMethodReference(Object function) {
+        final Method lambdaMethod = getCorrespondingMethod(function);
+
+        java.lang.reflect.Type genericReturnType = lambdaMethod.getGenericReturnType();
+        java.lang.reflect.Type genericParameterType = isStatic(lambdaMethod) ? lambdaMethod.getGenericParameterTypes()[0] : lambdaMethod.getDeclaringClass();
+
+        verifyNoRawType(genericReturnType);
+        verifyNoRawType(genericParameterType);
+
+        final Optional<TypeToken<R>> returnType = returnTypeToken(genericReturnType);
+        final TypeToken<T> argumentType = uncheckedTypeToken(genericParameterType);
+
+        return new FunctionInfo<>(returnType, argumentType);
+    }
+
+    private static Method getCorrespondingMethod(final Object functionalInterface) {
+        try {
+            final Method writeReplace = functionalInterface.getClass().getDeclaredMethod("writeReplace");
+            writeReplace.setAccessible(true);
+            final SerializedLambda lambda = (SerializedLambda) writeReplace.invoke(functionalInterface);
+
+            final String implMethodName = lambda.getImplMethodName();
+            final String implMethodSignature = lambda.getImplMethodSignature();
+
+            final Class<?> implClass = loadClass(lambda.getImplClass());
+
+            final Class<?>[] argumentClasses = Arrays.stream(Type.getArgumentTypes(implMethodSignature))
+                    .map(Type::getClassName)
+                    .map(TypeInference::loadClass)
+                    .toArray(Class<?>[]::new);
+
+            return implClass.getDeclaredMethod(implMethodName, argumentClasses);
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private static Class<?> loadClass(final String className) {
+        ClassLoader cl = Thread.currentThread().getContextClassLoader();
+        try {
+            return cl.loadClass(className.replace('/', '.'));
+        } catch (ClassNotFoundException e) {
+            throw new IllegalStateException("Could not load class [" + className + "]", e);
+        }
+    }
+
+    static <T, R> FunctionInfo<T, R> forFunction(final ThrowingFunction<T, R> function) {
+
+        if (function.getClass().isSynthetic()) {
+            return getFunctionInfoForMethodReference(function);
+        } else {
+            return getFunctionInfoForClass(function, ThrowingFunction.class);
+        }
+    }
+
+    static <T> FunctionInfo<T, Void> forConsumer(final ThrowingConsumer<T> function) {
+
+        if (function.getClass().isSynthetic()) {
```

Comment:
I need to create a synthetic class with custom `writeReplace` method to test code coverage here. I have no idea how to do that. Even with creating a class from bytecode, the flag seems to be cleared on class loading.